playbooks/preload-apt-newton.yml

```diff
@@ -0,0 +1,152 @@
+---
+# Preloads apt host packages for Newton OSA Release
+#
+# Can be run in advance of a leapfrog upgrade to Newton.
+# Allows for operator to prestage apt packages on hosts
+# to avoid bottleneck issues from retrieving packages
+# during the maintenance.
+# 
+# Useful for large number of hosts in an environment to
+# help reduce upgrade times.
+#
+# Pulls from vars for Ubuntu 14.04 LTS from
+# openstack-ansible-openstack_hosts
+# openstack-ansible-os_nova
+# openstack-ansible-lxc_host
+
+- name: Preflight check
+  hosts: hosts
+  gather_facts: false
+  user: root
+  tasks:
+    - name: Ping check
+      ping:
+
+- name: Ensure Apt is in a healthy state across all hosts
+  hosts: hosts
+  gather_facts: false
+  user: root
+  tasks:
+    - name: Ensure Apt is in a healthy state on all hosts
+      shell: "dpkg -a --configure; apt install -y -f"
+    - name: Refresh apt cache
+      apt:
+        update_cache: yes
+      ignore_errors: true
+
+- name: Preload openstack_host apt packages
+  hosts: hosts
+  user: root
+  vars:
+    openstack_host_distro_packages:
+      - apparmor-utils
+      - apt-transport-https
+      - bridge-utils
+      - build-essential
+      - cgroup-lite
+      - curl
+      - dmeventd
+      - dstat
+      - ebtables
+      - htop
+      - iptables
+      - irqbalance
+      - libkmod-dev
+      - libkmod2
+      - lvm2
+      - python-software-properties
+      - python-dev
+      - rsync
+      - rsyslog
+      - sshpass
+      - sysstat
+      - time
+      - vlan
+      - wget
+  tasks:
+    - name: Download apt packages but don't install
+      shell: "apt-get -d -y install {{ item }}"
+      with_items:
+        - "{{ openstack_host_distro_packages | join(' ') }}"
+
+- name: Preload lxc_hosts apt packages
+  hosts: lxc_hosts
+  user: root
+  vars:
+    lxc_hosts_distro_packages:
+      - apparmor
+      - apparmor-utils
+      - bridge-utils
+      - cgmanager
+      - cgroup-lite
+      - debootstrap
+      - dnsmasq-base
+      - git
+      - irqbalance
+      - language-pack-en
+      - liblxc1
+      - lxc
+      - lxc-dev
+      - lxc-templates
+      - python-dev
+      - python3-lxc
+      - pxz
+  tasks:
+    - name: Download apt packages but don't install
+      shell: "apt-get -d -y install {{ item }}"
+      with_items:
+        - "{{ lxc_hosts_distro_packages | join(' ') }}"
+
+- name: Preload compute_host apt packages
```

Comment:
This is a problem as the new cloud archives for newton are not laid down yet so the packaged you would download are actually not the ones you install

Reply:
It's mainly just doing a download only without the install (apt-get -d), so we could just add the repo, and then retrieve the files directly so they are sitting in the cache ready to go.  This could also just be done on the deployment host where we download all of the files into the cache and sync it around ahead of time.  In the larger host environments, it seems like this would be a useful operation to reduce time so that it's not chewing up time during a maintenance window.  That or set up an apt cache proxy.

Mainly just was kicking around ideas on how to avoid us slamming the mirror during the deployment... lots of things should be able to be done so that we can at least reduce the time of that step.  I think if we could just get something working for just the computes in large environments that would probably help a bit.

Reply:
Yepp I got that but the problem at this stage you are either missing the UCA or have older UCA repos deployed. This means you're not downloading the right package setup-hosts would do

Reply:
@BjoernT Made some tweaks, if you specify the target release, i.e newton and then it'll install the appropriate repo, download the files using apt, and the removes the repo added.  The files seem to remain in place though so if the repo is re-added they'll be there ready to use.

cd /opt/rpc-upgrades/playbooks
openstack-ansible preload-apt-packages.yml -e target_release=newton

It'll also let us define the vars based on release/os/version so that we can preload whatever is typically loaded during the upgrade.

Obviously, this doesn't help us on containers package gets but there are much fewer of those since they live on infras.  This is mainly geared toward large number of computes but it should help reduce the time we're stuck in the os-nova-install playbook.  I have some more tuning to do with newer versions, but it should be usable to test on Newton. (tested with Ansible 1.9.4, had to let it assume the filename for apt_repository, newer versions let you specify filename).